semantic_router/llms/base.py

```diff
@@ -20,81 +20,139 @@ def __call__(self, messages: List[Message]) -> Optional[str]:
+        self, query: str, function_schemas: list[dict[str, Any]]
     ) -> dict:
         logger.info("Extracting function input...")
 
         prompt = f"""
 You are an accurate and reliable computer program that only outputs valid JSON. 
-Your task is to output JSON representing the input arguments of a Python function.
+Your task is to:
+    1) Pick the most relevant Python function schema(s) from FUNCTION_SCHEMAS below, based on the input QUERY. If only one schema is provided, choose that. If multiple schemas are relevant, output a list of JSON objects for each.
+    2) Output JSON representing the input arguments of the chosen function schema(s), including the function name, with argument values determined by information in the QUERY.
 
-This is the Python function's schema:
+These are the Python functions' schema:
 
-### FUNCTION_SCHEMA Start ###
-	{function_schema}
-### FUNCTION_SCHEMA End ###
+### FUNCTION_SCHEMAS Start ###
+    {json.dumps(function_schemas, indent=4)}
+### FUNCTION_SCHEMAS End ###
 
 This is the input query.
 
 ### QUERY Start ###
-	{query}
+    {query}
 ### QUERY End ###
 
 The arguments that you need to provide values for, together with their datatypes, are stated in "signature" in the FUNCTION_SCHEMA.
 The values these arguments must take are made clear by the QUERY.
 Use the FUNCTION_SCHEMA "description" too, as this might provide helpful clues about the arguments and their values.
-Return only JSON, stating the argument names and their corresponding values.
+Include the function name in your JSON output.
+Return only JSON, stating the function name and the argument names with their corresponding values.
 
 ### FORMATTING_INSTRUCTIONS Start ###
-	Return a respones in valid JSON format. Do not return any other explanation or text, just the JSON.
-	The JSON-Keys are the names of the arguments, and JSON-values are the values those arguments should take.
+    Return a response in valid JSON format. Do not return any other explanation or text, just the JSON.
+    The JSON output should include a key 'function_name' with the value being the name of the function.
+    Under the key 'arguments', include a nested JSON object where the keys are the names of the arguments and the values are the values those arguments should take.
+    If multiple function schemas are relevant, return a list of JSON objects.
 ### FORMATTING_INSTRUCTIONS End ###
 
 ### EXAMPLE Start ###
-	=== EXAMPLE_INPUT_QUERY Start ===
-		"How is the weather in Hawaii right now in International units?"
-	=== EXAMPLE_INPUT_QUERY End ===
-	=== EXAMPLE_INPUT_SCHEMA Start ===
-		{{
-			"name": "get_weather",
-			"description": "Useful to get the weather in a specific location",
-			"signature": "(location: str, degree: str) -> str",
-			"output": "<class 'str'>",
-		}}
-	=== EXAMPLE_INPUT_QUERY End ===
-	=== EXAMPLE_OUTPUT Start ===
-		{{
-			"location": "Hawaii",
-			"degree": "Celsius",
-		}}
-	=== EXAMPLE_OUTPUT End ===
+    === EXAMPLE_INPUT_QUERY Start ===
```

Comment:
In the example, adding these extra tokens == additional wait time

It would be better to (1) see if other LLMs manage without the multiple examples, and (2) only use multiple examples if there are multiple function schemas present

Reply:
Both great ideas. will run testing without examples and if they're needed include functionality to swap out single examples for multiple examples. 

Should I also look into creating some sort of permanent testing scripts with which we can repeatedly asses performance moving forwards?

Reply:
yeah testing scripts seems like a good idea — let's keep that in a separate PR though